### DISCUSSION TITLE
FastAPI with Guvicorn is not able to receive SIGTERM signal when Kubernetes (K8S) pod scales down

### DISCUSSION THREAD

**Discussion by PranathiSG** [fastapi/fastapi#10609] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
shutdown_flag = False


def on_shutdown():
    # Handle cleanup or specific tasks when the application shuts down
    logger.info("Shutting down gracefully...")


@app.on_event("startup")
async def startup_event():
    def sigterm_handler(signum, frame):
        logger.info("Received SIGTERM. Initiating graceful shutdown.")
        loop = asyncio.get_event_loop()
        loop.create_task(wait_for_shutdown())

    signal.signal(signal.SIGTERM, sigterm_handler)


async def wait_for_shutdown():
    global shutdown_flag
    if shutdown_flag:
        return
    shutdown_flag = True

    try:
        # Perform cleanup or waiting for pending tasks here (if needed)
        # await asyncio.sleep(5)  # Simulated wait time
        await on_shutdown()
    except asyncio.CancelledError:
        pass


@app.on_event("shutdown")
async def shutdown_event():
    pass  # Do nothing during normal shutdown
```

### Description

FastAPI application is deployed in Kubernetes environment and when I try to scale down the application a SIGTERM signal is sent by the pods.

I've written the code in main.py file to handle SIGTERM signal but application is not receiving the SIGTERM and graceful shutdown is not working.

I've tested the code in docker container and it is working.

### Operating System

Linux

### FastAPI Version

0.75

### Pydantic Version

1.9

### Python Version

3.9

**Comment by jgould22** [comment, marked as answer]
How are you running the application ? 

With Gunicorn ?

With uvicorn ?

**Reply by vtgn** [reply]
@PranathiSG Are you serious?!!! Do you really think this is a good answer to the problem?!! #facepalm !!!

**Comment by iain17** [comment]
Gunicorn. Running into the same problem

**Comment by judgeaxl** [comment]
I've spent some time figuring out how to deal with this for uWSGI and gunicorn, and I imagine uvicorn or hypercorn can be dealt with in a s similar way:

First up, you want to set a pre-stop command, it does the following:

- Send a graceful shutdown command to gunicorn, then wait up to pre_stop_timeout seconds for gunicorn to stop.
- Once this finishes, Kubernetes will SIGTERM the container, which will brutally kill gunicorn.
- The initial sleep adds a buffer between k8s stops sending new requests to the pod, and we actually terminate them,
   "ensuring" that any request in flight will arrive and be considered during request draining.

```bash
/bin/bash -c sleep ${pre_stop_delay} && kill -s SIGTERM $(cat /tmp/gunicorn.pid) && timeout ${pre_stop_timeout} tail --pid=$(cat /tmp/gunicorn.pid) -f /dev/null
```

For the container command, do something like this:

- Wrap gunicorn with bash, so that when it terminates, we stick around for post_stop_delay seconds.
   Otherwise the pre_stop script will get killed the same instant gunicorn (PID 1) dies, resulting in an error in the logs.
```bash
/bin/bash -c gunicorn --config /app/config/gunicorn.conf.py; sleep ${post_stop_delay}
```

I use the following delays as I have a long-polling endpoint that times out after 20 seconds. If there are no polling requests currently in progress, it'll terminate roughly in `pre_stop_delay` + `post_stop_delay` seconds.

- `pre_stop_delay` = 3
- `pre_stop_timeout` = 30
- `post_stop_delay` = 2
- `pod_termination_grace_period` = 60

**Comment by o-fedorov** [comment]
I have the same issue with Kubernetes for both gunicorn and hypercorn. Similarly, everything seems to work fine for `docker run` and for the local running of the server.

Upon investigation, I found that for gunicorn, the issue arises when the command is baked into the Docker image using `CMD` or `ENTRYPOINT`. If the command is explicitly defined in a Kubernetes manifest with a `command` entry, SIGTERM seems to be handled correctly.

However, for hypercorn, even with the explicit definition of the command in Kubernetes manifest, the issue persists. According to the logs, it seems that out of the 3 workers spawned by hypercorn, only 2 stop gracefully, while one continues to work.

**Reply by o-fedorov** [reply]
Also, wrapping the command in a `sh` script solves the issue:

```yaml
      command:
        - sh
        - '-c'
        - 'hypercorn -c /path/to/hypercorn_conf.py src.app:app'
```

**Reply by alexey-nextreason** [reply]
Any updates on this issue? The solution mentioned by @o-fedorov doesn't work for me unfortunately :(
